specs/DesignDiscussions/DotNetTemplate.md

```diff
@@ -0,0 +1,338 @@
+```json
+{
+	"$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
+	"type": "AdaptiveCard",
+	"version": "1.0",
+	"body": [
+		{
+			"type": "TextBlock",
+			"text": "**wrap: true** {content}",
+			"wrap": true
+		}
+	]
+}
+```
+
+Notice that templateString node has been added, likewise other template keywords such as $data, $when, $root, $index can be parsed by the generated parser and lexer  
+Parser and lexer can be used on other platforms since they are generated by ANTLR tool and can be targeted to different languages. Visitor and Return Objects are needed to be implemented on other platforms.  
+  
+**Figure 3**
+![](antlr4_parse_tree_2.png)
+
+## Adaptive Expression Language
+This was initially intended to provide us functions support. AEL already has many built-in functions such as toLower, toString, equals, add, div, and [etc](https://github.com/microsoft/BotBuilder-Samples/blob/master/experimental/common-expression-language/prebuilt-functions.md)  
+But the AEL can do more. It can evaluate our templated string as well.
+
+```dotnet
+private ExpressionEngine _expressionEngine = new ExpressionEngine();
+/// some lines of code later
+var result = _expressionEngine.Parse(unboundString).TryEvaluate(data);
+```
+
+**Signature of TryEvaluate**
+```dotnet
+/// <summary>
+/// Evaluate the expression.
+/// </summary>
+/// <param name="state">
+/// Global state to evaluate accessor expressions against.  Can be <see cref="IDictionary{String}{Object}"/>, <see cref="IDictionary"/> otherwise reflection is used to access property and then indexer.
+/// </param>
+/// <returns>Computed value and an error string.  If the string is non-null, then there was an evaluation error.</returns>
+public (object value, string error) TryEvaluate(object state)
+```
+We just need to ensure our data implement IDictionary interface to be able to use this API for templated string expansion purposes.
+
+## Data Context
+- Data context is a data object adaptive element can use
+- Data context supports a memory concept, scope
+- Data context has two types, dictionary and array
+
+#### Scope
+Data objects can be placed anywhere in AdaptiveCards elements such as Container, TextBlock, and etc in json.
+Templating string is bound to the nearest data when AdaptiveCard is viewed as a tree
+
+```json
+{
+    "type": "AdaptiveCard",
+    "$data": {
+        "key": "1.2"
+    },
+    "body": [
+        {
+            "type": "Container",
+            "$data": {
+                "key": "2"
+            },
+            "items": [
+                {
+                    "type": "TextBlock",
+                    "$data": {
+                        "key": "3"
+                    },
+                    "text": "key should be 3, actual:{key}",
+                    "size": "Medium",
+                    "wrap": true
+                },
+                {
+                    "type": "TextBlock",
+                    "text": "key should be 2, actual:{key}",
```

Comment:
>key [](start = 54, length = 3)

Discussion: Should this have to be parent.key?

Reply:
Lots of discussion of the fact that the child nodes data context masks the entire data context of the parent. Some concern that we can't get to the parent's properties. This could be solved with support of an implicit parent.foo. This is not currently in plan but could theoretically be added to address this issue.

---
In reply to: [382840789](https://github.com/microsoft/AdaptiveCards/pull/3798#discussion_r382840789) [](ancestors = 382840789)

Reply:
I will consider it as p1 item, and re-evaluate it once all p0 items are complete.

---
In reply to: [382845909](https://github.com/microsoft/AdaptiveCards/pull/3798#discussion_r382845909) [](ancestors = 382845909,382840789)